File: server/sio/connection_manager.py
"""ConnectionManager — tracks active Socket.IO connections.

Provides add/remove/broadcast methods for managing WebSocket clients.
Uses python-socketio's room system for viewport-based broadcasting.
Includes heartbeat tracking and stale connection reaping.
"""

from __future__ import annotations

import asyncio
import logging
import time
from dataclasses import dataclass, field
from typing import Any

import socketio

logger = logging.getLogger(__name__)

# Heartbeat configuration
HEARTBEAT_INTERVAL_SEC = 10  # Server expects a heartbeat every 10s
HEARTBEAT_TIMEOUT_SEC = 30  # Client considered stale after 30s without heartbeat
REAPER_INTERVAL_SEC = 15  # How often the reaper checks for stale connections


@dataclass
class ConnectionInfo:
    """Metadata for a single connected client."""

    sid: str
    connected_at: float
    last_heartbeat: float = 0.0  # last app-level heartbeat timestamp
    rooms: set[str] = field(default_factory=set)
    user_id: str | None = None  # optional stable ID for reconnection
    reconnect_count: int = 0

    def __post_init__(self):
        if self.last_heartbeat == 0.0:
            self.last_heartbeat = self.connected_at


class ConnectionManager:
    """Manages active Socket.IO connections.

    Thread-safe for asyncio — all mutations happen in the event loop.

    Features:
    - Connection add/remove lifecycle
    - Heartbeat tracking with stale connection reaping
    - Room-based viewport subscriptions
    - Reconnection state restoration via user_id
    - Broadcast to all, to rooms, or to individual clients

    Usage:
        sio = socketio.AsyncServer(async_mode="asgi")
        manager = ConnectionManager(sio)

        # On connect
        await manager.add(sid)

        # On subscribe:viewport — join grid rooms
        await manager.join_rooms(sid, grid_ids)

        # Broadcast fire update to a specific room
        await manager.broadcast_to_room("fire:update", data, room=grid_id)

        # Broadcast to all connected clients
        await manager.broadcast("fire:update", data)

        # On disconnect
        await manager.remove(sid)
    """

    def __init__(self, sio: socketio.AsyncServer) -> None:
        self._sio = sio
        self._connections: dict[str, ConnectionInfo] = {}
        # Map user_id → last known ConnectionInfo (for reconnection)
        self._user_sessions: dict[str, ConnectionInfo] = {}
        self._reaper_task: asyncio.Task | None = None

    @property
    def active_count(self) -> int:
        """Number of currently active connections."""
        return len(self._connections)

    @property
    def active_sids(self) -> list[str]:
        """List of all active session IDs."""
        return list(self._connections.keys())

    def get(self, sid: str) -> ConnectionInfo | None:
        """Get connection info for a given sid."""
        return self._connections.get(sid)

    async def add(
        self, sid: str, user_id: str | None = None
    ) -> ConnectionInfo:
        """Register a new connection.

        Args:
            sid: Socket.IO session ID.
            user_id: Optional stable user identifier for reconnection support.

        Returns:
            ConnectionInfo for the newly added connection.
        """
        now = time.time()
        reconnect_count = 0

        # Check if this user_id has a previous session (reconnection)
        if user_id and user_id in self._user_sessions:
            prev = self._user_sessions[user_id]
            reconnect_count = prev.reconnect_count + 1
            logger.info(
                "Reconnection detected: user=%s prev_sid=%s new_sid=%s count=%d",
                user_id, prev.sid, sid, reconnect_count,
            )

        info = ConnectionInfo(
            sid=sid,
            connected_at=now,
            last_heartbeat=now,
            user_id=user_id,
            reconnect_count=reconnect_count,
        )
        self._connections[sid] = info

        if user_id:
            self._user_sessions[user_id] = info

        logger.info("Connection added: %s (total: %d)", sid, self.active_count)
        return info

    async def remove(self, sid: str) -> ConnectionInfo | None:
        """Unregister a connection and leave all rooms.

        Args:
            sid: Socket.IO session ID.

        Returns:
            ConnectionInfo of the removed connection, or None if not found.
        """
        info = self._connections.pop(sid, None)
        if info is None:
            logger.warning("Attempted to remove unknown sid: %s", sid)
            return None

        # Leave all tracked rooms
        for room in info.rooms:
            try:
                await self._sio.leave_room(sid, room)
            except Exception:
                pass  # sid may already be disconnected

        logger.info("Connection removed: %s (total: %d)", sid, self.active_count)
        return info

    async def join_rooms(self, sid: str, rooms: list[str]) -> None:
        """Subscribe a client to one or more rooms (grid IDs).

        Replaces any previously subscribed rooms — a client only watches
        one viewport at a time.

        Args:
            sid: Socket.IO session ID.
            rooms: List of room names (typically grid IDs) to join.
        """
        info = self._connections.get(sid)
        if info is None:
            logger.warning("join_rooms called for unknown sid: %s", sid)
            return

        # Leave old rooms
        for old_room in info.rooms:
            try:
                await self._sio.leave_room(sid, old_room)
            except Exception:
                pass

        # Join new rooms
        new_rooms: set[str] = set()
        for room in rooms:
            await self._sio.enter_room(sid, room)
            new_rooms.add(room)

        info.rooms = new_rooms
        logger.debug("sid %s joined %d rooms", sid, len(new_rooms))

    async def broadcast(self, event: str, data: Any) -> None:
        """Broadcast an event to ALL connected clients.

        Args:
            event: Socket.IO event name (e.g. "fire:update").
            data: Payload to send.
        """
        await self._sio.emit(event, data)
        logger.debug("Broadcast '%s' to all (%d clients)", event, self.active_count)

    async def broadcast_to_room(
        self, event: str, data: Any, room: str
    ) -> None:
        """Broadcast an event to all clients in a specific room.

        Args:
            event: Socket.IO event name.
            data: Payload to send.
            room: Room name (typically a grid ID).
        """
        await self._sio.emit(event, data, room=room)
        logger.debug("Broadcast '%s' to room '%s'", event, room)

    async def send_to(self, event: str, data: Any, sid: str) -> None:
        """Send an event to a specific client.

        Args:
            event: Socket.IO event name.
            data: Payload to send.
            sid: Target client's session ID.
        """
        await self._sio.emit(event, data, to=sid)

    # ------------------------------------------------------------------
    # Heartbeat
    # ------------------------------------------------------------------

    def record_heartbeat(self, sid: str) -> bool:
        """Record a heartbeat from the client.

        Args:
            sid: Socket.IO session ID.

        Returns:
            True if the connection exists and heartbeat was recorded.
        """
        info = self._connections.get(sid)
        if info is None:
            return False
        info.last_heartbeat = time.time()
        return True

    def is_stale(self, sid: str, now: float | None = None) -> bool:
        """Check if a connection has missed heartbeats.

        Args:
            sid: Socket.IO session ID.
            now: Current timestamp (defaults to time.time()).

        Returns:
            True if the connection is stale (no heartbeat within timeout).
        """
        info = self._connections.get(sid)
        if info is None:
            return True
        if now is None:
            now = time.time()
        return (now - info.last_heartbeat) > HEARTBEAT_TIMEOUT_SEC

    def get_stale_sids(self, now: float | None = None) -> list[str]:
        """Get all sids that have exceeded the heartbeat timeout.

        Args:
            now: Current timestamp (defaults to time.time()).

        Returns:
            List of stale session IDs.
        """
        if now is None:
            now = time.time()
        return [
            sid
            for sid, info in self._connections.items()
            if (now - info.last_heartbeat) > HEARTBEAT_TIMEOUT_SEC
        ]

    # ------------------------------------------------------------------
    # Stale connection reaper
    # ------------------------------------------------------------------

    def start_reaper(self) -> None:
        """Start the background task that disconnects stale clients."""
        if self._reaper_task is not None:
            return
        self._reaper_task = asyncio.create_task(self._reaper_loop())
        logger.info("Stale connection reaper started (interval=%ds)", REAPER_INTERVAL_SEC)

    async def stop_reaper(self) -> None:
        """Stop the stale connection reaper."""
        if self._reaper_task is None:
            return
        self._reaper_task.cancel()
        try:
            await self._reaper_task
        except asyncio.CancelledError:
            pass
        self._reaper_task = None
        logger.info("Stale connection reaper stopped")

    async def _reaper_loop(self) -> None:
        """Periodically check and disconnect stale connections."""
        while True:
            try:
                await asyncio.sleep(REAPER_INTERVAL_SEC)
                stale = self.get_stale_sids()
                for sid in stale:
                    logger.warning(
                        "Reaping stale connection: %s", sid
                    )
                    try:
                        await self._sio.disconnect(sid)
                    except Exception:
                        # Force-remove from our tracking if disconnect fails
                        await self.remove(sid)
                if stale:
                    logger.info("Reaped %d stale connection(s)", len(stale))
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("Error in reaper loop")

    # ------------------------------------------------------------------
    # Reconnection support
    # ------------------------------------------------------------------

    def get_previous_session(self, user_id: str) -> ConnectionInfo | None:
        """Get the previous session info for a user_id.

        Used during reconnection to restore room subscriptions.

        Args:
            user_id: Stable user identifier.

        Returns:
            Previous ConnectionInfo or None.
        """
        return self._user_sessions.get(user_id)

    async def restore_rooms(self, sid: str, rooms: set[str]) -> None:
        """Restore room subscriptions for a reconnected client.

        Args:
            sid: New Socket.IO session ID.
            rooms: Set of room names to rejoin.
        """
        if not rooms:
            return
        await self.join_rooms(sid, list(rooms))

File: server/internal/sio/connection_manager.go
// Package sio provides application-level Socket.IO connection management.
//
// It mirrors server/sio/connection_manager.py — tracking active connections,
// heartbeat state, room subscriptions, and reconnection support via user_id.
package sio

import (
	"log"
	"sync"
	"time"
)

// Heartbeat / reaper configuration — matches Python server values.
const (
	HeartbeatIntervalSec = 10 // server expects a heartbeat every 10s
	HeartbeatTimeoutSec  = 30 // client considered stale after 30s
	ReaperIntervalSec    = 15 // reaper check interval
)

// ConnectionInfo holds metadata for a single connected client.
// Mirrors Python's ConnectionInfo dataclass.
type ConnectionInfo struct {
	SID            string
	ConnectedAt    float64 // Unix timestamp (seconds)
	LastHeartbeat  float64
	Rooms          map[string]struct{} // set of room names (grid IDs)
	UserID         string              // optional stable ID for reconnection
	ReconnectCount int
}

// newConnectionInfo creates a ConnectionInfo with sensible defaults.
func newConnectionInfo(sid string, userID string, now float64, reconnectCount int) *ConnectionInfo {
	return &ConnectionInfo{
		SID:            sid,
		ConnectedAt:    now,
		LastHeartbeat:  now,
		Rooms:          make(map[string]struct{}),
		UserID:         userID,
		ReconnectCount: reconnectCount,
	}
}

// RoomNames returns a copy of the room set as a string slice.
func (ci *ConnectionInfo) RoomNames() []string {
	rooms := make([]string, 0, len(ci.Rooms))
	for r := range ci.Rooms {
		rooms = append(rooms, r)
	}
	return rooms
}

// ConnectionManager tracks active Socket.IO connections.
//
// Thread-safe: all exported methods are protected by a mutex.
//
// Features (matching Python implementation):
//   - Connection add/remove lifecycle
//   - Heartbeat tracking with stale connection detection
//   - Room-based viewport subscriptions
//   - Reconnection state restoration via user_id
type ConnectionManager struct {
	mu            sync.RWMutex
	connections  map[string]*ConnectionInfo // sid → ConnectionInfo
	userSessions map[string]*ConnectionInfo // user_id → last known ConnectionInfo
	logger       *log.Logger
}

// NewConnectionManager creates a new ConnectionManager.
func NewConnectionManager(logger *log.Logger) *ConnectionManager {
	if logger == nil {
		logger = log.Default()
	}
	return &ConnectionManager{
		connections:  make(map[string]*ConnectionInfo),
		userSessions: make(map[string]*ConnectionInfo),
		logger:       logger,
	}
}

// ActiveCount returns the number of currently active connections.
func (m *ConnectionManager) ActiveCount() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.connections)
}

// ActiveSIDs returns a copy of all active session IDs.
func (m *ConnectionManager) ActiveSIDs() []string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	sids := make([]string, 0, len(m.connections))
	for sid := range m.connections {
		sids = append(sids, sid)
	}
	return sids
}

// Get returns the ConnectionInfo for a given sid, or nil if not found.
func (m *ConnectionManager) Get(sid string) *ConnectionInfo {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.connections[sid]
}

// Add registers a new connection. If the user_id was seen before,
// increments the reconnect count (matching Python behavior).
//
// Returns the newly created ConnectionInfo.
func (m *ConnectionManager) Add(sid string, userID string) *ConnectionInfo {
	now := float64(time.Now().UnixMilli()) / 1000.0 // seconds with ms precision
	reconnectCount := 0

	m.mu.Lock()
	defer m.mu.Unlock()

	// Check for previous session (reconnection detection)
	if userID != "" {
		if prev, ok := m.userSessions[userID]; ok {
			reconnectCount = prev.ReconnectCount + 1
			m.logger.Printf("Reconnection detected: user=%s prev_sid=%s new_sid=%s count=%d",
				userID, prev.SID, sid, reconnectCount)
		}
	}

	info := newConnectionInfo(sid, userID, now, reconnectCount)
	m.connections[sid] = info

	if userID != "" {
		m.userSessions[userID] = info
	}

	m.logger.Printf("Connection added: %s (total: %d)", sid, len(m.connections))
	return info
}

// Remove unregisters a connection and returns its ConnectionInfo.
// Returns nil if the sid was not found.
// Note: room cleanup in the Socket.IO namespace is the caller's responsibility.
func (m *ConnectionManager) Remove(sid string) *ConnectionInfo {
	m.mu.Lock()
	defer m.mu.Unlock()

	info, ok := m.connections[sid]
	if !ok {
		m.logger.Printf("Attempted to remove unknown sid: %s", sid)
		return nil
	}

	delete(m.connections, sid)

	if info.UserID != "" {
		if current, ok := m.userSessions[info.UserID]; ok && current.SID == sid {
			delete(m.userSessions, info.UserID)
		}
	}

	m.logger.Printf("Connection removed: %s (total: %d)", sid, len(m.connections))
	return info
}

// GetPreviousSession returns the last known session for a user_id.
// Used during reconnection to restore room subscriptions.
func (m *ConnectionManager) GetPreviousSession(userID string) *ConnectionInfo {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.userSessions[userID]
}

// RecordHeartbeat updates the last heartbeat timestamp for a sid.
// Returns true if the connection exists.
func (m *ConnectionManager) RecordHeartbeat(sid string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	info, ok := m.connections[sid]
	if !ok {
		return false
	}
	info.LastHeartbeat = float64(time.Now().UnixMilli()) / 1000.0
	return true
}

// IsStale checks if a connection has missed heartbeats.
func (m *ConnectionManager) IsStale(sid string) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	info, ok := m.connections[sid]
	if !ok {
		return true
	}
	now := float64(time.Now().UnixMilli()) / 1000.0
	return (now - info.LastHeartbeat) > HeartbeatTimeoutSec
}

// GetStaleSIDs returns all sids that have exceeded the heartbeat timeout.
func (m *ConnectionManager) GetStaleSIDs() []string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	now := float64(time.Now().UnixMilli()) / 1000.0
	var stale []string
	for sid, info := range m.connections {
		if (now - info.LastHeartbeat) > HeartbeatTimeoutSec {
			stale = append(stale, sid)
		}
	}
	return stale
}

// SetRooms replaces the tracked rooms for a sid.
// This only updates the ConnectionManager's internal tracking;
// actual Socket.IO room join/leave is the caller's responsibility.
func (m *ConnectionManager) SetRooms(sid string, rooms []string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	info, ok := m.connections[sid]
	if !ok {
		return
	}
	info.Rooms = make(map[string]struct{}, len(rooms))
	for _, r := range rooms {
		info.Rooms[r] = struct{}{}
	}
}

// GetRooms returns the current rooms for a sid.
func (m *ConnectionManager) GetRooms(sid string) []string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	info, ok := m.connections[sid]
	if !ok {
		return nil
	}
	return info.RoomNames()
}

